peers.rb

```diff
@@ -17,18 +18,148 @@
+  elsif password.size == 0
+    puts "Please enter a valid email password"
+    erb :signup_test, layout: :footer_layout
+  elsif full_name.size == 0
+    puts "Please enter a valid full name"
+    erb :signup_test, layout: :footer_layout
+  else
+    password = encrypt_password(password)
+    @storage.add_new_user(email, password, full_name)
+    puts "Your account has been created. Please login using your credentials"
+    redirect "/login"
+  end
+end
+
+get "/login" do
+  erb :login_test, layout: :footer_layout
+end
+
+post "/login" do
+  email = params[:email]
+  password = params[:password]
+
+  if valid_credentials?(email, password)
+    @user = @storage.get_user_by_email(email)
+    puts "Login completed"
+    redirect "/profile/#{@user.id}"
+  else
+    puts "Sorry, your credentials don't exist in our database. Please try again or create an account."
+    erb :login_test, layout: :footer_layout
+  end
+end
+
+get "/logout" do         # To be implemented
+  # So far it just redirects to "/". I tried with '@storage.disconnet' but raises an error
+  puts "Logout complete"
+  redirect "/"
+end
+
+get "/profile/:id" do
+  user_id = params[:id]
+  @user = @storage.get_user_by_id(user_id)
+  if is_user_login?(@user.id, @user.email)               # implemented as:  id == @storage.get_id_by_email(email)
+    erb :profile, layout: :footer_layout
+  else
+    puts "Sorry, you must be login to view this content"
+    erb :login_test, layout: :footer_layout
+  end
+end
+
+get "/profile/:id/edit" do
+  user_id = params[:id]
+  @user = @storage.get_user_by_id(user_id)
+  erb :profile_edit_test, layout: :footer_layout
 end
 
-get "/test/profile/:id" do
+post "/profile/:id/edit" do                # Shall we make this a PUT route instead (see Notions comments on routes section)
   user_id = params[:id]
-  @user = @storage.user_by_id(user_id)
+  preferred_name = params[:preferred_name]
+  slack_name = params[:slack_name]
+  track = params[:track]
+  course = params[:course]
+  timezone = params[:timezone]
+  preferences = params[:preferences]
+  about_me = params[:about_me]
 
-  erb :profile, layout: :footer_layout
+  if preferred_name.size == 0
+    puts "Please enter a valid preferred name"
+    erb :profile_edit_test, layout: :footer_layout
+  elsif slack_name.size == 0
+    puts "Please enter a valid slack name"
+    erb :profile_edit_test, layout: :footer_layout
+  elsif track.size == 0
+    puts "Please enter a valid track"
+    erb :profile_edit_test, layout: :footer_layout
+  elsif course.size == 0
+    puts "Please enter a valid course"
+    erb :profile_edit_test, layout: :footer_layout
+  elsif timezone.size == 0
+    puts "Please enter a valid timezone"
+    erb :profile_edit_test, layout: :footer_layout
+  elsif preferences.size == 0
+    puts "Please enter a valid preference"
+    erb :profile_edit_test, layout: :footer_layout
+  else
+    @storage.update_user_data(user_id, preferred_name, slack_name, track, course, timezone, about_me)
```

Comment:
`@storage.update_user_data` can only handle foreign key values for track, course, timezone and preferences. This is fine for now, but we'll need to make sure we do a mapping so that users can input values that they understand (e.g. 'RB' for track rather than '1')